Brand's customers

arijit1111 opened the pull request:
Brand's customers

arijit1111 committed:
feedbacks on feature/customer pull

arijit1111 committed:
Merge remote-tracking branch 'origin/feature/customer' into feature/customer

# Conflicts:
#	Modules/Customer/Http/Services/CustomerService.php

arijit1111 committed:
changes according frontend need

smaqeelkazmi commented on Modules/Customer/Http/Controllers/CustomerController.php:
```diff
@@ -121,26 +121,25 @@ public function update(UpdateCustomerRequest $request): JsonResponse
         }
 
-        $response = $this->customerService->delete($customerKey);
+        $response = $this->customerService->delete($request->all());
```
@arijit1111 I believe it should be the customer ID, which we are deleting from DB, instead of whole request object

smaqeelkazmi replied:
@arijit1111 Also, specifically get the parameter of list of the customers instead of the whole request object.

arijit1111 replied:
It is only passing customers id array and validation added in DeleteCustomerRequest

smaqeelkazmi commented on Modules/Customer/Http/Services/CustomerService.php:
```diff
@@ -117,18 +121,18 @@ public function update(array $requestData): array
-            $notSignedCustomersCount = Customer::where('brand_id', $brand->user_id)->where('status', 'not signed up')->count();
-            $onBazarCustomersCount = Customer::where('brand_id', $brand->user_id)->where('status', 'on bazar')->count();
-            $customers = Customer::where('brand_id', $brand->user_id);
+            $allCustomersCount = Customer::where('user_id', $brand->user_id)->count();
```
@arijit1111 You can change it to the local scope as `Customer::auth()` In local scope the method should check for authorized user if the instance belongs to.

smaqeelkazmi replied:
@arijit1111 Aren't you going to make this change right now?

arijit1111 replied:
These are updated in last push

arijit1111 replied:
updated in latest push

smaqeelkazmi commented on Modules/Customer/Http/Services/CustomerService.php:
```diff
@@ -199,24 +203,23 @@ public function getCustomers(Request $request): array
      * @return array
      */
-    public function delete(string $customerKey): array
+    public function delete(array $requestData): array
```
@arijit1111 The list of customer IDs should be passed through the body of the request instead of URL path

smaqeelkazmi reviewed (changes requested):
I have added few comments, kindly check them

arijit1111 committed:
customers feedbacks fixed

smaqeelkazmi commented on Modules/Customer/Database/Migrations/2023_03_04_141823_create_customers_table.php:
```diff
@@ -15,10 +15,21 @@ public function up()
         Schema::create('customers', function (Blueprint $table) {
             $table->id();
             $table->unsignedBigInteger('user_id')->nullable();
+            $table->unsignedBigInteger('retailer_id')->nullable();
```
@arijit1111 Kindly create new migration with update table action. Then add all of the required changes in the existing database table over there. Kindly revert all of these changes.

smaqeelkazmi commented on Modules/Customer/Entities/Customer.php:
```diff
@@ -14,13 +14,31 @@ class Customer extends Model {
+        'shipping_town',
+        'shipping_zip',
+        'shipping_phone_code',
+        'shipping_phone',
```
@arijit1111 The `retailed_id` is missing in fillable

smaqeelkazmi commented on Modules/Customer/Entities/Customer.php:
```diff
@@ -14,13 +14,31 @@ class Customer extends Model {
+     */
+    public function scopeAuth(Builder $query, string $user_id): void
+    {
+        $query->where('user_id', $user_id);
```
@arijit1111 Kindly change this to `$query->where('user_id', Auth::id());`

smaqeelkazmi replied:
@arijit1111 I cannot see this change

smaqeelkazmi reviewed (changes requested):
Kindly review the request changes

arijit1111 committed:
revert back all the changes in create customers table migration
create new migration for update customers table and add the required changes
update local scope for customers fetched by auth user
added retailer_id to fillable

smaqeelkazmi commented on Modules/Customer/Http/Services/CustomerService.php:
```diff
@@ -100,6 +120,44 @@ public function update(array $requestData): array
+            ];
+        }
+
         $customer->update($requestData);
```
@arijit1111 We are validating few of the parameters, but what if I pass a parameter other than the validated one? Lets say we are validating A,B,C but what if user passes D which haven't been validated through the Request while the column exists in Database, it'll directly update in the database and is harmful to security.

arijit1111 replied:
validation added in UpdateCustomerInfoRequest, UpdateCustomerShippingRequest

smaqeelkazmi commented on Modules/Customer/Http/Controllers/CustomerController.php:
```diff
@@ -121,26 +170,75 @@ public function update(UpdateCustomerRequest $request): JsonResponse
+        $user = auth()->user();
+
+        // return error if user is not a brand
+        if ($user->cannot('viewAny', Customer::class)) {
```
@arijit1111 This should be `delete` Policy Method instead of `viewAny`

arijit1111 replied:
It is only to check is the user brand or not, it is a list of customers id in request and now it validated as customers array

arijit1111 replied:
it is only to check whether user is brand or not

smaqeelkazmi reviewed (changes requested):
@arijit1111 I have added two more changes request, kindly check them.

smaqeelkazmi commented on Modules/Customer/Database/Migrations/2023_05_23_130502_update_customers_table.php:
```diff
@@ -0,0 +1,44 @@
+     */
+    public function down()
+    {
+
```
@arijit1111 N o migration effect have been defined for `down` migration action

arijit1111 replied:
added in last push

arijit1111 committed:
down method is now updated for update customers table migration

smaqeelkazmi reviewed (approved):
PR approved
